File: 114.二叉树展开为链表.cpp
/*
 * @lc app=leetcode.cn id=114 lang=cpp
 *
 * [114] 二叉树展开为链表
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */
class Solution {
  private:
    TreeNode *cur;

  public:
    void flatten(TreeNode *root) {
        if (!root)
            return;
        preorder(root);
    }

    TreeNode *preorder(TreeNode *root) {
        if (!root)
            return root;
        if (cur) {
            cur->left = nullptr;
            cur->right = root;
            cur = cur->right;
        } else {
            cur = root;
        }
        auto tp = cur->right;
        auto tp2 = cur->left;

        preorder(tp2);
        preorder(tp);
        return root;
    }
};
// @lc code=end

File: 114.二叉树展开为链表.go
/*
 * @lc app=leetcode.cn id=114 lang=golang
 *
 * [114] 二叉树展开为链表
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

 func flatten(root *TreeNode) {
	if root == nil {
		return
	}
	var f func(root *TreeNode) *TreeNode
	f = func(root *TreeNode) *TreeNode {
		if root == nil || root.Left == nil && root.Right == nil {
			return root
		}
		ret := f(root.Left)
		root.Left = nil
		if ret != nil {
			right := root.Right
			root.Right = ret
			for ret.Right != nil {
				ret = ret.Right
			}
			ret.Right = f(right)
		} else {
			root.Right = f(root.Right)
		}
		return root
	}
	f(root)
}
// @lc code=end

